### PR TITLE
HBASE-26045 Master control the global throughput of all compaction servers

### PR DESCRIPTION
**Pull request by nyl3532016**
Compaction servers report compaction tasks to  master via heartbeat report. 
The report response has throughput control message and compaction server use this message to adjust compaction throughtput.
We can set total throughtput of all compaction servers via hbase shell commands.

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   2m  9s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +0 :ok: |  prototool  |   0m  0s |  prototool was not available.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ HBASE-25714 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 27s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   5m 21s |  HBASE-25714 passed  |
| +1 :green_heart: |  compile  |   9m  9s |  HBASE-25714 passed  |
| +1 :green_heart: |  checkstyle  |   3m 33s |  HBASE-25714 passed  |
| +1 :green_heart: |  spotbugs  |  10m 40s |  HBASE-25714 passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 12s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   4m  2s |  the patch passed  |
| +1 :green_heart: |  compile  |   7m  7s |  the patch passed  |
| +1 :green_heart: |  cc  |   7m  7s |  the patch passed  |
| +1 :green_heart: |  javac  |   7m  7s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 28s |  hbase-client: The patch generated 1 new + 38 unchanged - 0 fixed = 39 total (was 38)  |
| -0 :warning: |  checkstyle  |   1m 11s |  hbase-server: The patch generated 7 new + 6 unchanged - 0 fixed = 13 total (was 6)  |
| -0 :warning: |  rubocop  |   0m 29s |  The patch generated 20 new + 656 unchanged - 0 fixed = 676 total (was 656)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |  20m 24s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.1 3.3.0.  |
| +1 :green_heart: |  hbaseprotoc  |   2m 57s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |  10m 12s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 53s |  The patch does not generate ASF License warnings.  |
|  |   |  89m 57s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/3538 |
| Optional Tests | dupname asflicense javac spotbugs hadoopcheck hbaseanti checkstyle compile cc hbaseprotoc prototool rubocop |
| uname | Linux 6bcc747bc9d2 4.15.0-147-generic #151-Ubuntu SMP Fri Jun 18 19:21:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-25714 / 85f02919da |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/1/artifact/yetus-general-check/output/diff-checkstyle-hbase-client.txt |
| checkstyle | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/1/artifact/yetus-general-check/output/diff-checkstyle-hbase-server.txt |
| rubocop | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/1/artifact/yetus-general-check/output/diff-patch-rubocop.txt |
| Max. process+thread count | 86 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-client hbase-server hbase-thrift hbase-shell U: . |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/1/console |
| versions | git=2.17.1 maven=3.6.3 spotbugs=4.2.2 rubocop=0.80.0 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 34s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  5s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ HBASE-25714 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 23s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   4m 40s |  HBASE-25714 passed  |
| +1 :green_heart: |  compile  |   3m 55s |  HBASE-25714 passed  |
| +1 :green_heart: |  shadedjars  |   8m 22s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   2m 27s |  HBASE-25714 passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   4m 24s |  the patch passed  |
| +1 :green_heart: |  compile  |   3m 48s |  the patch passed  |
| +1 :green_heart: |  javac  |   3m 48s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   8m 10s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   2m 27s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m  1s |  hbase-protocol-shaded in the patch passed.  |
| +1 :green_heart: |  unit  |   1m 18s |  hbase-client in the patch passed.  |
| +1 :green_heart: |  unit  | 142m 19s |  hbase-server in the patch passed.  |
| +1 :green_heart: |  unit  |   4m 40s |  hbase-thrift in the patch passed.  |
| +1 :green_heart: |  unit  |   7m 17s |  hbase-shell in the patch passed.  |
|  |   | 199m  9s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/1/artifact/yetus-jdk11-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/3538 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux f83a3a0d36ed 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-25714 / 85f02919da |
| Default Java | AdoptOpenJDK-11.0.10+9 |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/1/testReport/ |
| Max. process+thread count | 3694 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-client hbase-server hbase-thrift hbase-shell U: . |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 35s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  5s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ HBASE-25714 Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 22s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   4m 25s |  HBASE-25714 passed  |
| +1 :green_heart: |  compile  |   3m 28s |  HBASE-25714 passed  |
| +1 :green_heart: |  shadedjars  |   8m 29s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   2m  7s |  HBASE-25714 passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   4m  5s |  the patch passed  |
| +1 :green_heart: |  compile  |   3m 29s |  the patch passed  |
| +1 :green_heart: |  javac  |   3m 29s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   9m 22s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   2m  9s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 52s |  hbase-protocol-shaded in the patch passed.  |
| +1 :green_heart: |  unit  |   1m 18s |  hbase-client in the patch passed.  |
| -1 :x: |  unit  | 155m 15s |  hbase-server in the patch failed.  |
| +1 :green_heart: |  unit  |   4m 14s |  hbase-thrift in the patch passed.  |
| +1 :green_heart: |  unit  |   7m 26s |  hbase-shell in the patch passed.  |
|  |   | 211m 11s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/1/artifact/yetus-jdk8-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/3538 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux 5d39b79c2fbc 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | HBASE-25714 / 85f02919da |
| Default Java | AdoptOpenJDK-1.8.0_282-b08 |
| unit | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/1/artifact/yetus-jdk8-hadoop3-check/output/patch-unit-hbase-server.txt |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/1/testReport/ |
| Max. process+thread count | 3760 (vs. ulimit of 30000) |
| modules | C: hbase-protocol-shaded hbase-client hbase-server hbase-thrift hbase-shell U: . |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-client/src/main/java/org/apache/hadoop/hbase/client/Admin.java [hunk 1]
```diff
@@ -2256,6 +2256,17 @@ void cloneTableSchema(TableName tableName, TableName newTableName, boolean prese
    */
   boolean isCompactionOffloadEnabled() throws IOException;
 
+  /**
+   * update compaction server total throughput bound
+   * @param upperBound the total throughput upper bound of all compaction servers
```
Comment (saintstack) [inline hbase-client/src/main/java/org/apache/hadoop/hbase/client/Admin.java:2261]: When we talk of compaction servers here, are we talking about the new compaction server facility or are we talking about regionservers?

Comment (nyl3532016) [reply hbase-client/src/main/java/org/apache/hadoop/hbase/client/Admin.java:2261]: we can control compaction throughtput on compactionServers, can not control regionservers

Comment (saintstack) [reply hbase-client/src/main/java/org/apache/hadoop/hbase/client/Admin.java:2261]: Most deploys will not use a compaction server suite. Is it correct then to talk of compaction servers in our Admin API? Should there be another channel for interacting with compaction servers? A CompactionServersAdmin ?

Comment (nyl3532016) [reply hbase-client/src/main/java/org/apache/hadoop/hbase/client/Admin.java:2261]: This method talk to master only, compaction servers get throughput control message through periodic heartbeat report

**Comment by nyl3532016** [comment]
@Apache9 @saintstack Mind help take a look at your convenience.

**Review by Apache9 — COMMENTED** [review]
So the algorithm here does not count for ongoing compactions, instead, just divide the quota evenly across different compaction servers? If there are 10 compaction server, but only one compaction server has ongoing compactions, it still can only use 1/10 of the quota?

**Comment by nyl3532016** [comment]
> So the algorithm here does not count for ongoing compactions, instead, just divide the quota evenly across different compaction servers? If there are 10 compaction server, but only one compaction server has ongoing compactions, it still can only use 1/10 of the quota?

OK, let me promote the algorithm according to ongoing compactions

File: hbase-client/src/main/java/org/apache/hadoop/hbase/client/Admin.java [hunk 2]
```diff
@@ -2256,6 +2256,17 @@ void cloneTableSchema(TableName tableName, TableName newTableName, boolean prese
...
+   * @param lowerBound the total throughput lower bound of all compaction servers
+   * @param offPeak the total throughput offPeak bound of all compaction servers
+   * @return the now total throughput of all compaction servers
+   * @throws IOException if a remote or network exception occurs
+   */
+  Map<String, Long> updateCompactionServerTotalThroughput(Long upperBound, Long lowerBound,
```
[the remark this answers is not in the document]
Comment (nyl3532016) [reply hbase-client/src/main/java/org/apache/hadoop/hbase/client/Admin.java:2267]: Yes, We do not control regionserver's compaction throughtput in this PR,  For HRegionserver can adjust their compaction throughtput via update_config shell commend, no need restart server. They implement ConfigurationObserver interface.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m  0s |  Docker mode activated.  |
| -1 :x: |  patch  |   0m  5s |  https://github.com/apache/hbase/pull/3538 does not apply to HBASE-25714. Rebase required? Wrong Branch? See https://yetus.apache.org/documentation/in-progress/precommit-patchnames for help.  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hbase/pull/3538 |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/2/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m  0s |  Docker mode activated.  |
| -1 :x: |  patch  |   0m  5s |  https://github.com/apache/hbase/pull/3538 does not apply to HBASE-25714. Rebase required? Wrong Branch? See https://yetus.apache.org/documentation/in-progress/precommit-patchnames for help.  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hbase/pull/3538 |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/2/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m  0s |  Docker mode activated.  |
| -1 :x: |  patch  |   0m  5s |  https://github.com/apache/hbase/pull/3538 does not apply to HBASE-25714. Rebase required? Wrong Branch? See https://yetus.apache.org/documentation/in-progress/precommit-patchnames for help.  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hbase/pull/3538 |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-3538/2/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Pull request closed without merging**